firedrake/assemble.py

```diff
@@ -545,80 +518,124 @@ def get_map(x):
+        "nest", "aij", "baij", or "matfree".
+    :arg sub_mat_type: (optional) type for assembled sub matrices
+        inside a "nest" matrix.  One of "aij" or "baij".
+    :arg appctx: Additional information to hang on the assembled
+         matrix if an implicit matrix is requested (mat_type "matfree").
+    :arg options_prefix: An options prefix for the PETSc matrix
+        (ignored if not assembling a bilinear form).
+    """
+    mat_type, sub_mat_type = get_mat_type(mat_type, sub_mat_type)
+    if form_compiler_parameters:
+        form_compiler_parameters = form_compiler_parameters.copy()
+    else:
+        form_compiler_parameters = {}
+
+    try:
+        topology, = set(d.topology for d in f.ufl_domains())
+    except ValueError:
+        raise NotImplementedError("All integration domains must share a mesh topology")
+    for m in f.ufl_domains():
+        # Ensure mesh is "initialised" (could have got here without
+        # building a functionspace (e.g. if integrating a constant)).
+        m.init()
+
+    if bcs is None:
+        bcs = ()
+    else:
+        bcs = tuple(bcs)
+
+    for o in chain(f.arguments(), f.coefficients()):
+        domain = o.ufl_domain()
+        if domain is not None and domain.topology != topology:
+            raise NotImplementedError("Assembly with multiple meshes not supported.")
+
+    rank = len(f.arguments())
+    if diagonal:
+        assert rank == 2
+    if rank == 2 and not diagonal:
+        assembly_rank = AssemblyRank.MATRIX
+    elif rank == 1 or diagonal:
+        assembly_rank = AssemblyRank.VECTOR
+    else:
+        assembly_rank = AssemblyRank.SCALAR
+
+    if assembly_rank == AssemblyRank.MATRIX:
+        test, trial = f.arguments()
+        tensor, zeros, result = get_matrix(f, mat_type, sub_mat_type,
+                                           bcs=bcs, options_prefix=options_prefix,
+                                           tensor=tensor,
+                                           appctx=appctx,
+                                           form_compiler_parameters=form_compiler_parameters)
+        # intercept matrix-free matrices here
+        if mat_type == "matfree":
+            if tensor.a.arguments() != f.arguments():
+                raise ValueError("Form's arguments do not match provided result "
+                                 "tensor")
+            tensor.assemble()
+            yield result
+            return
+
+        create_op2arg = functools.partial(matrix_arg,
+                                          all_bcs=tuple(chain(*bcs)),
+                                          tensor=tensor,
+                                          Vrow=test.function_space(),
+                                          Vcol=trial.function_space())
+    elif assembly_rank == AssemblyRank.VECTOR:
+        if diagonal:
+            # actually a 2-form but throw away the trial space
+            test, trial = f.arguments()
```

Comment:
Might be a good idea to make sure the test and trial spaces are equal.

Reply:
Done.